src/box/rt_access.c

```diff
@@ -0,0 +1,111 @@
+#include "rt_access.h"
+#include "assoc.h"
+#include "../lua/init.h" /* tarantool_lua_is_builtin_global */
+#include "salad/grp_alloc.h"
+
+static struct mh_strn_strn_t *user_rt_access = NULL;
+static struct mh_strnptr_t *user_rt_universe_access = NULL;
+
+/**
+ * Helper structure for grp_alloc.
+ */
+struct user_name {
+	/* User name. */
+	const char *str;
+	/* User name len. */
+	uint32_t len;
+};
+
+void
+rt_access_init(void)
+{
+	user_rt_access = mh_strn_strn_new();
+	user_rt_universe_access = mh_strnptr_new();
+}
+
+void
+rt_access_free(void)
+{
+	mh_int_t i;
+	if (user_rt_access != NULL) {
+		mh_foreach(user_rt_access, i) {
+			struct mh_strn_strn_node_t *node =
+				mh_strn_strn_node(user_rt_access, i);
+			free(node->val);
+		}
+	}
+	if (user_rt_universe_access != NULL) {
+		mh_foreach(user_rt_universe_access, i) {
+			struct mh_strnptr_node_t *node =
+				mh_strnptr_node(user_rt_universe_access, i);
+			free(node->val);
+		}
+	}
+}
+
+void
+rt_access_destroy(void)
+{
+	rt_access_free();
+	mh_strn_strn_delete(user_rt_access);
+	mh_strnptr_delete(user_rt_universe_access);
+}
+
+void
+rt_access_reset(void)
+{
+	rt_access_free();
+	mh_strn_strn_clear(user_rt_access);
+	mh_strnptr_clear(user_rt_universe_access);
+}
+
+void
+rt_access_grant(const char *uname, uint32_t uname_len,
+		const char *fname, uint32_t fname_len)
+{
+	struct user_name *node;
+	struct grp_alloc all = grp_alloc_initializer();
+	grp_alloc_reserve_data(&all, sizeof(*node));
+	grp_alloc_reserve_str(&all, uname_len);
+	grp_alloc_use(&all, xmalloc(grp_alloc_size(&all)));
+	node = (struct user_name *)grp_alloc_create_data(&all, sizeof(*node));
+	node->str = grp_alloc_create_str(&all, uname, uname_len);
+	node->len = uname_len;
+	assert(grp_alloc_size(&all) == 0);
+
+	/* Grant universe access. */
+	if (fname_len == 0) {
+		uint32_t hash = mh_strn_hash(node->str, uname_len);
+		const struct mh_strnptr_node_t uid_node =
+			{ node->str, uname_len, hash, node };
+		mh_strnptr_put(user_rt_universe_access, &uid_node, NULL, NULL);
```

Comment:
If the node exists, it is replaced by the new node. It seems, we loss the memory allocated by `grp_alloc` above in the case.

The same for `mh_strn_strn_put()` below.

Reply:
Fixed.

Reply:
`rt_access_check` may return `true` if the permission is granted for universe, while the `rt_access_grant` is called for a particular function.

Since there is no revoke function it works valid, but I would prefer a solution that would skip putting a value in the hashmap only if the value actually present in *this* hashmap.

I think that it would be less error-prone this way.